stores/migrations.go

```diff
@@ -645,6 +665,71 @@ func performMigration00013_uploadPackingOptimisations(txn *gorm.DB, logger *zap.
 			return err
 		}
 	}
-	logger.Info(context.Background(), "migration 00013_uploadPackingOptimisations complete")
+	logger.Info("migration 00013_uploadPackingOptimisations complete")
+	return nil
+}
+
+func performMigration00014_buckets(txn *gorm.DB, logger *zap.SugaredLogger) error {
+	logger.Info("performing migration 00014_buckets")
+	// Disable foreign keys in SQLite to avoid issues with updating constraints.
+	if isSQLite(txn) {
+		if err := txn.Exec(`PRAGMA foreign_keys = 0`).Error; err != nil {
+			return err
+		}
+	}
+
+	// Create buckets table
+	if !txn.Migrator().HasTable(&dbBucket{}) {
+		if err := txn.Migrator().CreateTable(&dbBucket{}); err != nil {
+			return err
+		}
+	}
+
+	// Add default bucket.
+	bucket := &dbBucket{
+		Name: api.DefaultBucketName,
+	}
+	if err := txn.FirstOrCreate(&bucket).Error; err != nil {
+		return err
+	}
+
+	// Add bucket id column to objects table.
+	if !txn.Migrator().HasColumn(&dbObject{}, "db_bucket_id") {
+		if err := txn.Migrator().AddColumn(&dbObject{}, "db_bucket_id"); err != nil {
+			return err
+		}
+	}
+
+	// Create missing composite index.
+	if !txn.Migrator().HasIndex(&dbObject{}, "idx_object_bucket") {
+		if err := txn.Migrator().CreateIndex(&dbObject{}, "idx_object_bucket"); err != nil {
+			return err
+		}
+	}
+
+	// Update objects to belong to default bucket
+	if err := txn.Model(&dbObject{}).
+		Where("db_bucket_id", 0).
+		Update("db_bucket_id", bucket.ID).Error; err != nil {
+		return err
+	}
+
+	// Add foreign key constraint between dbObject's db_bucket_id and dbBucket's id.
+	if !txn.Migrator().HasConstraint(&dbObject{}, "DBBucket") {
+		if err := txn.Migrator().CreateConstraint(&dbObject{}, "DBBucket"); err != nil {
+			return err
+		}
+	}
+
+	// Enable foreign keys again.
+	if isSQLite(txn) {
+		if err := txn.Exec(`PRAGMA foreign_keys = 1`).Error; err != nil {
+			return err
+		}
+		if err := txn.Exec(`PRAGMA foreign_key_check(slabs)`).Error; err != nil {
```

Comment:
we're explicitly checking slabs and not objects? is that copy pasta or intentional?